### PR TITLE
[exporter/fileexporter] Support for compressing the telemetry data before exporting

### PR DESCRIPTION
**Pull request by atingchen**
Description:
Support for compressing the telemetry data before exporting

Link to tracking Issue:
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13626

Testing:
Updated unit tests

Documentation:
Update Readme.md

### CODE REVIEW

File: exporter/fileexporter/file_exporter.go [hunk 1]
```diff
@@ -46,6 +47,8 @@ var logsMarshalers = map[string]plog.Marshaler{
 	formatTypeProto: plog.NewProtoMarshaler(),
 }
 
+var encoder, _ = zstd.NewWriter(nil)
```
Comment (djaglowski) [inline exporter/fileexporter/file_exporter.go:50]: Are you sure we can use a single instance of the encoder?

It's possible that a single instance of this exporter will receive data from many pipelines. Would this be a problem?

File: exporter/fileexporter/README.md [hunk 2]
```diff
@@ -33,7 +35,7 @@ The following settings are optional:
   - localtime : [default: false (use UTC)] whether or not the timestamps in backup files is formatted according to the host's local time.
 
 - `format`[default: json]: define the data format of encoded telemetry data. The setting can be overridden with `proto`.
-
+- `compression`[default:false]:defines whether to compress encoded telemetry data.
```
Comment (djaglowski) [inline exporter/fileexporter/README.md:38]: We may eventually want to support other compression algorithms. Instead of using a boolean, why don't we use a string? The default can be an empty string, and this will imply no compression. `"zstd"` can act the same as you currently have for `true`.

File: exporter/fileexporter/file_exporter.go (hunk 1 above)
Comment (atingchen) [reply exporter/fileexporter/file_exporter.go:50]: The [README](https://github.com/klauspost/compress/tree/master/zstd) indicates this function `EncodeAll` can be called concurrently. Each call will only run on a same goroutine as the caller.
At the same time, I added unit test `TestConcurrentlyCompress` to check whether using the `Encoder` for individual blocks concurrently is safe.
So, Can we reuse the encoder?

Comment (atingchen) [reply exporter/fileexporter/file_exporter.go:50]: I will save the encoder on the exporter struct,just in case.

File: exporter/fileexporter/README.md (hunk 2 above)
Comment (atingchen) [reply exporter/fileexporter/README.md:38]: Thanks for your advice～

File: exporter/fileexporter/README.md [hunk 3]
```diff
@@ -33,8 +34,7 @@ The following settings are optional:
   - localtime : [default: false (use UTC)] whether or not the timestamps in backup files is formatted according to the host's local time.
 
 - `format`[default: json]: define the data format of encoded telemetry data. The setting can be overridden with `proto`.
-
-
+- `compression`[default = none]: the compression algorithm used when exporting telemetry data to file. The options are: `none` and `zstd`
```
Comment (djaglowski) [inline exporter/fileexporter/README.md:37]:
```suggestion
- `compression`[no default]: the compression algorithm used when exporting telemetry data to file. Supported compression algorithms:`zstd`
```

File: exporter/fileexporter/config.go [hunk 4]
```diff
@@ -67,8 +71,11 @@ func (cfg *Config) Validate() error {
 	if cfg.Path == "" {
 		return errors.New("path must be non-empty")
 	}
-	if cfg.FormatType != formatTypeJSON && cfg.FormatType != formatTypeProto {
+	if cfg.FormatType != defaultFormatType && cfg.FormatType != formatTypeProto {
 		return errors.New("format type is not supported")
 	}
+	if cfg.Compression != defaultCompression && cfg.Compression != compressionZSTD {
```
Comment (djaglowski) [inline exporter/fileexporter/config.go:77]:
```suggestion
	if cfg.Compression != "" && cfg.Compression != compressionZSTD {
```

File: exporter/fileexporter/file_exporter.go [hunk 5]
```diff
@@ -27,23 +27,18 @@ import (
...
-	formatTypeJSON:  pmetric.NewJSONMarshaler(),
-	formatTypeProto: pmetric.NewProtoMarshaler(),
+	defaultFormatType: pmetric.NewJSONMarshaler(),
+	formatTypeProto:   pmetric.NewProtoMarshaler(),
 }
 var logsMarshalers = map[string]plog.Marshaler{
-	formatTypeJSON:  plog.NewJSONMarshaler(),
-	formatTypeProto: plog.NewProtoMarshaler(),
+	defaultFormatType: plog.NewJSONMarshaler(),
+	formatTypeProto:   plog.NewProtoMarshaler(),
```
Comment (djaglowski) [inline exporter/fileexporter/file_exporter.go:41]: Unrelated

File: exporter/fileexporter/factory.go [hunk 6]
```diff
@@ -30,8 +30,17 @@ const (
...
+
 	// the number of old log files to retain
 	defaultMaxBackups = 100
+
+	// the format of encoded telemetry data
+	defaultFormatType = "json"
+	formatTypeProto   = "proto"
+
+	// the type of compression codec
+	defaultCompression = "none"
```
Comment (djaglowski) [inline exporter/fileexporter/factory.go:42]: No need to have a value for "none". Just work with `""`.
```suggestion
```

File: exporter/fileexporter/file_exporter.go [hunk 7]
```diff
@@ -129,7 +130,7 @@ func (e *fileExporter) Shutdown(context.Context) error {
 }
 
 func buildExportFunc(cfg *Config) func(e *fileExporter, buf []byte) error {
-	if cfg.FormatType == formatTypeProto {
+	if cfg.FormatType == formatTypeProto || cfg.Compression != defaultCompression {
```
Comment (djaglowski) [inline exporter/fileexporter/file_exporter.go:133]: I am unclear about the relationship between `format` and `compression` here. 

For example, why does `json && zstd` behave the same as `proto && none`?

Comment (atingchen) [reply exporter/fileexporter/file_exporter.go:133]: It's really not easy for other to understand here, I'll fix it

File: exporter/fileexporter/README.md [hunk 8]
```diff
@@ -46,16 +46,20 @@ in the name immediately before the file's extension (or the end of the filename
...
-##  File Format
+##  File Format 
 
 Telemetry data is encoded according to the `format` setting and then written to the file.
 
-When `format` is json, telemetry data is written to file in JSON format. Each line in the file is a JSON object.
-When `format` is proto, each encoded object is preceded by 4 bytes (an unsigned 32 bit integer) which represent the number of bytes contained in the encoded object.
+When `format` is json and `compression` is none , telemetry data is written to file in JSON format. Each line in the file is a JSON object.
+
+Besides, each encoded object is preceded by 4 bytes (an unsigned 32 bit integer) which represent the number of bytes contained in the encoded object.When we need read the messages back in, we read the size, then read the bytes into a separate buffer, then parse from that buffer.
```
Comment (djaglowski) [inline exporter/fileexporter/README.md:61]:
```suggestion
Otherwise, when using `proto` format or any kind of encoding, each encoded object is preceded by 4 bytes (an unsigned 32 bit integer) which represent the number of bytes contained in the encoded object.When we need read the messages back in, we read the size, then read the bytes into a separate buffer, then parse from that buffer.
```

Comment (atingchen) [reply exporter/fileexporter/README.md:61]: Thanks

**Review by djaglowski — APPROVED** [review]
Thanks @atingchen
